File: src/main/java/upbrella/be/user/validation/OnlyNumbers.java
package upbrella.be.user.validation;

import javax.validation.Constraint;
import javax.validation.Payload;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Target({ElementType.FIELD, ElementType.METHOD, ElementType.PARAMETER, ElementType.ANNOTATION_TYPE})
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = OnlyNumbersValidator.class)
public @interface OnlyNumbers {
    String message() default "The field should contain only numbers";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: src/main/kotlin/upbrella/be/user/validation/OnlyNumbers.kt
package upbrella.be.user.validation

import javax.validation.Constraint
import javax.validation.Payload
import kotlin.reflect.KClass

@Target(AnnotationTarget.FIELD, AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER, AnnotationTarget.VALUE_PARAMETER, AnnotationTarget.ANNOTATION_CLASS)
@Retention(AnnotationRetention.RUNTIME)
@Constraint(validatedBy = [OnlyNumbersValidator::class])
annotation class OnlyNumbers(
    val message: String = "The field should contain only numbers",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)